### PR TITLE
Improved the selection of subflows for the default scheduler

### PR DESCRIPTION
**Pull request by ghost**
I reorganized the code as you suggested and tested it. It seems to work as expected.

### CODE REVIEW

File: net/mptcp/mptcp_sched.c [hunk 1]
```diff
@@ -88,11 +116,28 @@ static bool mptcp_is_available(struct sock *sk, const struct sk_buff *skb,
 	 */
 	if (skb && !zero_wnd_test &&
 	    after(tp->write_seq + min(skb->len, mss_now), tcp_wnd_end(tp)))
+		return true;
+
+	return false;
+}
+
+/* If the sub-socket sk available to send the skb? */
```
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:124]: s/If/Is :)

File: net/mptcp/mptcp_sched.c [hunk 2]
```diff
@@ -104,6 +149,64 @@ static int mptcp_dont_reinject_skb(const struct tcp_sock *tp, const struct sk_bu
...
 }
 
+static bool subflow_is_selected(struct tcp_sock *tp, struct sk_buff *skb,
+				enum sk_groups sk_group)
+{
+	switch (sk_group) {
+	case ACTIVE_NOT_USED:
+		if (!subflow_is_backup(tp) && !mptcp_dont_reinject_skb(tp, skb))
+			return true;
+		return false;
```
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:159]: You could simply do:
return !subflow_is_backup(tp) && !mptcp_dont_reinject_skb(tp, skb);

Instead of the if-statement. The same for the other if-clauses.

**Comment by cpaasch** [comment]
Hello,

thanks for updating your patch :)

Some high-level comments:

Please merge the two patches in one single patch.

Also, I would prefer if you merge the subflow-selection for active_not_used and active_used into one single call to get_subflow_from_group. Because, with your changes we will always iterate twice over the full list of subflows, if their congestion-windows are full (which is always the case with bulk-transfers).

Also, why not using the function-pointer in get_subflow_from_group()? This would allow to avoid using the switch-selection.

I added some detailled comments inline.

File: net/mptcp/mptcp_sched.c [hunk 3]
```diff
@@ -15,58 +22,79 @@ static struct defsched_priv *defsched_get_priv(const struct tcp_sock *tp)
...
-	if (!mptcp_sk_can_send(sk))
-		return false;
+	/* Set of TCP states for which we are not allowed to send data
+	 * "temporally"
+	 */
+	if (sk->sk_state == TCPF_SYN_SENT && !tcp_passive_fastopen(sk))
+		return true;
+
+	if ((1 << sk->sk_state) &
+	    (TCPF_SYN_RECV | TCPF_TIME_WAIT | TCPF_LAST_ACK))
```
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:57]: I'm unsure, why time-wait and last-ack are in the "temporal" set.
Also, syn-recv should in my opinion also go to the definitive unavailable set.

File: net/mptcp/mptcp_sched.c [hunk 4]
```diff
@@ -88,11 +116,28 @@ static bool mptcp_is_available(struct sock *sk, const struct sk_buff *skb,
...
+}
+
+/* If the sub-socket sk available to send the skb? */
+static bool mptcp_is_available(struct sock *sk, const struct sk_buff *skb,
+			       bool zero_wnd_test)
+{
+	if (mptcp_is_def_unavailable(sk) ||
+	    mptcp_is_temp_unavailable(sk, skb, zero_wnd_test)) {
 		return false;
+	}
```
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:131]: Here you can also simply do:
return !mptcp_is_def_unavailable() && !mptcp_is_temp_unavailable();

File: net/mptcp/mptcp_sched.c (hunk 1 above)
Comment (ghost) [inline net/mptcp/mptcp_sched.c:124]: +1

File: net/mptcp/mptcp_sched.c (hunk 2 above)
Comment (ghost) [inline net/mptcp/mptcp_sched.c:159]: +1

File: net/mptcp/mptcp_sched.c (hunk 4 above)
Comment (ghost) [inline net/mptcp/mptcp_sched.c:131]: +1

**Comment by ghost** [comment]
Hello Christoph,
- I will provide you a final one commit with all changes but until obtain the final code I prefer to continue providing new commits, it is easier to manage for me.
- We need a comprimise between the number of iterations and the number of variables. With the initial approach we only need one iteration over the full sk list but we need four groups of variables (srrt and sk pointers for: bestsk, used_bestsk, lowpriosk and used_lowpriosk). With the current approach we need four iterations but we only need one group of variables (srrt and sk pointer). As you say the 2 iterators might be a compromise.
- I tought about your preference for the selector function and now :) I understand that the selector function perhaps is easier to mantain because it is easier to add/delete selector functions instead of modifying continously the same function with the switch clause? 

So, I propose the following skeleton for the get_subflow_from_group renamed as get_subflow_from_selectors

``` C
static struct sock *get_subflow_from_selectors(struct mptcp_cb *mpcb,
                       struct sk_buff *skb,
                       bool (*main_selector)(struct tcp_sock *, struct sk_buff *),
                                           bool (*backup_selector)(struct tcp_sock *, struct sk_buff *),
                       bool zero_wnd_test,
                                           bool *force)
{
        cnt = 0
        cnt_temp_una = 0
        cnt_def_una = 0

        mptcp_for_each_sk(mpcb, sk) {
        tp = tcp_sk(sk);
        if ((*main_selector)(tp, skb))
            main = true;
                else if ((*backup_selector)(tp, skb))
                        main = false;
                else
                        continue;
        cnt++;
        if (mptcp_is_def_unavailable(sk)) {
            cnt_def_una++;
            continue;
        }
        if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test)) {
            cnt_temp_una++;
            continue;
        }
        if (main && tp->srtt_us < main_min_srtt) {
            main_min_srtt = tp->srtt_us;
            main_sk = sk;
        }
                if (!main && tp->srtt_us < backup_min_srtt) {
            backup_min_srtt = tp->srtt_us;
            backup_sk = sk;
        }
    }

        if (main_sk)
            *force_ = true
            return main_sk
        else if (backup_sk)
             * force = true
             TCP_SKB_CB(skb)->path_mask = 0;
             return backup_sk
        else if (cnt != 0 && cnt != cnt_def_una)
              * force = true
              return NULL
         else
               *force = false
               return NULL
}
```

File: net/mptcp/mptcp_sched.c (hunk 3 above)
Comment (ghost) [inline net/mptcp/mptcp_sched.c:57]: Yes perhaps you are right, taking a look at the TCP Connection State Diagram of the RFC 793 it seems that TIME_WAIT and LAST_ACK will finish in a connection closed and can't return to the ESTAB state. So they should we considered as "definetively" unavailable.

However, the SYN_RECV under normal circunstances will quickly pass to the ESTAB state, so In my opinion it should be in the "temporal" unavaible set.

Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:57]: Concerning SYN_RECV, in general you should never have a socket in SYN_RECV-state that is not a fastopen-socket. Because, Linux creates request-sockets upon reception of a SYN, which are not exposed as TCP-sockets.

File: net/mptcp/mptcp_sched.c [hunk 5]
```diff
@@ -15,58 +22,79 @@ static struct defsched_priv *defsched_get_priv(const struct tcp_sock *tp)
...
 	unsigned int mss_now, space, in_flight;
 
-	/* Set of states for which we are allowed to send data */
-	if (!mptcp_sk_can_send(sk))
-		return false;
+	/* Set of TCP states for which we are not allowed to send data
+	 * "temporally"
+	 */
+	if (sk->sk_state == TCPF_SYN_SENT && !tcp_passive_fastopen(sk))
+		return true;
```
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:54]: A socket in SYN_SENT that is not a fastopen-socket might hang around quite some time before it becomes established (depends on the RTT). So, I would also classify SYN_SENT as definitely unavailable.

File: net/mptcp/mptcp_sched.c (hunk 3 above)
Comment (ghost) [inline net/mptcp/mptcp_sched.c:57]: SYN_RECV state is reached when a SYN is received by the host and once the three way-handsake is finished with the reception of the ACK, the TCP connection passes to ESTAB state. Perhaps, TCP-sockets are not externally exposed but I understand that this state must be managed by TCP-stack to reach the ESTAB state.

In any case, in mptcp_ctrl.c and mptcp_output.c there are also checks of the TCPF_SYN_RECV state.

Perhaps, the safest procedure is to check also this state just in case. If so, I would check it in the temporal unavaible-set instead of the definetive unavaible-set.

Comment (ghost) [inline net/mptcp/mptcp_sched.c:57]: To pass from TCPF_SYN_RECV to TCPF_ESTAB, 1 RTT is required (the server sends the SYN-ACK and must receive the ACK prior to send any data package). However, MPTCP requires more ACKs to fully establish the connection so 2 RTT are required to have a MPTCP-ESTAB connection. Yes perhaps, it is too high and it is better to use the backup.

**Comment by cpaasch** [comment]
Hello,

> On Jan 8, 2015, at 2:03 AM, chrpinedo notifications@github.com wrote:
>   • I will provide you a final one commit with all changes but until obtain the final code I prefer to continue providing new commits, it is easier to manage for me.

ok, fair enough.

>   • We need a comprimise between the number of iterations and the number of variables. With the initial approach we only need one iteration over the full sk list but we need four groups of variables (srrt and sk pointers for: bestsk, used_bestsk, lowpriosk and used_lowpriosk). With the current approach we need four iterations but we only need one group of variables (srrt and sk pointer). As you say the 2 iterators might be a compromise.

I think you can design it differently. Because, as soon as you find one non-backup subflow that is eligible, you can stop tracking backup-subflows. So, I think you might be able to drop the second variables rtt-variable for the backupsk.

>   • I tought about your preference for the selector function and now :) I understand that the selector function perhaps is easier to mantain because it is easier to add/delete selector functions instead of modifying continously the same function with the switch clause?
> 
> So, I propose the following skeleton for the get_subflow_from_group renamed as get_subflow_from_selectors

I am guessing that with "backup_selector" you mean the one counting those subflows who would reinject skb's?

> static struct sock _get_subflow_from_selectors(struct mptcp_cb *mpcb, struct sk_buff *skb, bool (_main_selector)(struct tcp_sock _, struct sk_buff *), bool (_backup_selector)(struct tcp_sock *, struct sk_buff *), bool zero_wnd_test, bool *force)
> {
>         cnt = 0
>         cnt_temp_una = 0
>         cnt_def_una = 0
> 
>   mptcp_for_each_sk (mpcb, sk) {
>           tp = tcp_sk(sk);
>       if ((_main_selector)(tp, skb))
>                   main = true;
>       else if ((_backup_selector)(tp, skb))

So, here you could do "else if (!cnt_temp_una && backup_selector(...))"

> ```
>                     main = false;
>   else
>       continue;
>   cnt++;
> 
>   if (mptcp_is_def_unavailable(sk)) {
>       cnt_def_una++;
>               continue;
>       }
> 
>   if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test)) {
> ```

I think, you need to split this test here for backup subflows.

> ```
>       cnt_temp_una++;
>       continue;
>   }
> 
>   if (main && tp->srtt_us < main_min_srtt) {
>       main_min_srtt = tp->srtt_us;
>       main_sk = sk;
>   }
> ```

This, would become:
if (tp->srtt_us < min_srtt) {
    min_srtt = tp->srtt_us;
    bestsk = sk;
}

> ```
>   if (!main && tp->srtt_us < backup_min_srtt) {
>       backup_min_srtt = tp->srtt_us;
>       backup_sk = sk;
>   }
> ```

And you would entirely drop this if-clause.

>   }
> 
>   if (main_sk) {
>       *force_ = true;
> 
> ```
>   return main_sk;
> ```
> 
>   } else if (backup_sk) {
>       *force = true;
> 
> ```
>   return backup_sk;
> ```
> 
>   } else if (cnt != 0 && cnt != cnt_def_una) {
>       *force = true;
> 
> ```
>   return NULL;
> ```
> 
>   } else {
>                *force = false;
> 
> ```
>   return NULL;
> ```
> 
>   }

I have the feeling that the following is correct:

static bool main_selector(tp, skb)
{
    return !tp->mptcp->low_prio && !tp->mptcp_rcv_lowprio;
}

static struct sock _get_subflow_from_selectors(struct mptcp_cb *mpcb,
                           struct sk_buff *skb, bool (_main_selector)(struct tcp_sock *, struct sk_buff *),
                           bool zero_wnd_test, bool *force)
{
        cnt = 0
        cnt_temp_una = 0
        cnt_def_una = 0
    bool found_unused = false;
    bestsk = NULL;

```
mptcp_for_each_sk (mpcb, sk) {
    bool unused = false;
        tp = tcp_sk(sk);

    /* First, we choose among the main-set - whether this is a low-prio subflow or not */
    if (!(*main_selector)(tp, skb))
        continue;

    if (!mptcp_dont_reinject_skb())
                    unused = true;
    else if (found_unused)
        /* If this is a reinjection, and we already had an unused subflow that
         * was eligible, we continue - no need to check reinjection-subflows anymore.
         */
        continue;

    if (mptcp_is_def_unavailable(sk))
                continue;

    if (unused) {
        /* It's the first time we encounter an unused
         * subflow - thus we reset the bestsk (which might
         * have been set to a reinjection-subflow.
         */
        if (!found_unused)
            bestsk = NULL;
        found_unused = true;
    }
    if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test))
        continue;

    if (tp->srtt_us < main_min_srtt) {
        min_srtt = tp->srtt_us;
        bestsk = sk;
    }

}

if (found_unused)
    *force = true

return bestsk;
```

}

File: net/mptcp/mptcp_sched.c (hunk 3 above)
Comment (cpaasch) [inline net/mptcp/mptcp_sched.c:57]: Yes, please move SYN_RECV to the definitely unavailable set.

**Comment by ghost** [comment]
buff, really good approach but it was difficult for me to understand it :-D

Differences:
- you can send a NULL sk even when there is an used active sk that can be selected to send  the skb where there is no unused active sk. In my opinion it is best to send the skb through the used active sk and no wait for an unused active sk. (This is different to used/unused backup sks). So we should move the "if (unused)" block after the "if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test))" block.
- if we use a used sk we should clear the skb's path_mask inside this function or outside.
- I don't have enough information to know if we should call this function a second time for the backup sks. I only need to iterate over the backup sks if: "found_unused" is false, "found_used" is false and there are active sk (used/unused) and all of them are def unavailable. If there used/unused active sk that are temp unavailable, we should wait for them and not search for backup sks.

**Comment by ghost** [comment]
Lest's clarify **which states we have to move to def unavaible function or temp unavailable function**. To consider it we can take a look at the following function you are calling today from mptcp_is_available:

``` C
static inline int mptcp_sk_can_send(const struct sock *sk)
{
    return tcp_passive_fastopen(sk) ||
           ((1 << sk->sk_state) & (TCPF_ESTABLISHED | TCPF_CLOSE_WAIT) &&
        !tcp_sk(sk)->mptcp->pre_established);
}
```

blocks temporally:
- TCPF_ESTABLISHED && tcp_sk(sk)->mptcp->pre_established (we are not able to send but soon we will be able to)
- TCPF_CLOSE_WAIT && tcp_sk(sk)->mptcp->pre_established **or** TCPF_CLOSE_WAIT **??**  (**Doubt!** according to mptcp_sk_can_send function it is possible to send data in this state if no in pre_established MPTCP state. We remove it? or maintain it to avoid data sending in this state? def or temp unavailability?)

block definetivelly:
- TCPF_SYN_RECV && !tcp_passive_fastopen(sk) (too many time to begin to send, if fast open not the case. In fast open in this state we can send data in fact)
- TCPF_SYN_SENT (too many time to begin to send)
- TCPF_FIN_WAIT1
- TCPF_FIN_WAIT2
- TCPF_TIME_WAIT
- TCPF_CLOSE
- TCPF_CLOSE_WAIT && tcp_sk(sk)->mptcp->pre_established **or** TCPF_CLOSE_WAIT **??**   (**Doubt!** according to mptcp_sk_can_send function it is possible to send data in this state if no in pre_established MPTCP state. We remove it? or maintain it to avoid data sending in this state? def or temp unavailability?)
- TCPF_LAST_ACK
- TCPF_LISTEN
- TCPF_CLOSING

**Comment by cpaasch** [comment]
> On Jan 8, 2015, at 9:37 AM, chrpinedo notifications@github.com wrote:
> 
> buff, really good approach but it was difficult for me to understand it :-D
> 
> Differences:
> 
>   • you can send a NULL sk even when there is an used active sk that can be selected to send the skb where there is no unused active sk. In my opinion it is best to send the skb through the used active sk and no wait for an unused active sk. (This is different to used/unused backup sks). So we should move the "if (unused)" block after the "if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test))" block.

Then we would be continuously reinjecting segments on used sk's when we the pipes (congestion windows) of the other sk's are all filled.

As an example, what will happen is the following:

Subflow A, subflow B. On subflow A we have sent segment 1, on subflow B we have sent segment 2, 3 and 4. Subflow B's congestion window is full.

If now the scheduler selects to resend segment 1, we really do not want to send it yet again on subflow A. We want to first send it on subflow B and thus want to wait until subflow B has again space in its congestion window.

These reinjections on "used" subflows should be very rare. Because, after all, the segment has already been sent on the "used" subflow once and thus should have reached the receiver. It should only be necessary when facing a segment coalescing middlebox. Thus, we must make sure that we do not waste bandwidth by forcing these reinjections too often.

>   • if we use a used sk we should clear the skb's path_mask inside this function or outside.

Indeed.

>   • I don't have enough information to know if we should call this function a second time for the backup sks. I only need to iterate over the backup sks if: "found_unused" is false, "found_used" is false and there are active sk (used/unused) and all of them are def unavailable. If there used/unused active sk that are temp unavailable, we should wait for them and not search for backup sks.

After calling this function, you should probably do something like the following:

```
sk = get_subflow_from_selectors(..., &main_selector,...)
if (force || sk)
    return sk;


/* Now look for backup-sk's */
sk = get_subflow_from_selectors(... &backup_selector,...)
...
```

**Comment by cpaasch** [comment]
> On Jan 8, 2015, at 9:56 AM, chrpinedo notifications@github.com wrote:
> 
> Lest's clarify which states we have to move to def unavaible function or temp unavailable function. To consider it we can take a look at the following function you are calling today from mptcp_is_available:
> 
> static inline int mptcp_sk_can_send(const struct
>  sock *sk)
> {
> 
> return tcp_passive_fastopen
> (sk) ||
>            ((
> 1
>  << sk->sk_state) & (TCPF_ESTABLISHED | TCPF_CLOSE_WAIT) &&
>         !
> tcp_sk
> (sk)->mptcp->pre_established);
> }
> 
> blocks temporally:
> 
>   • TCPF_ESTABLISHED && tcp_sk(sk)->mptcp->pre_established (we are not able to send but soon we will be able to)

pre-established state may not necessarily be a temporal state. The subflow may be broken and it takes quite a long time until the retransmission-timer and exponential backoff has reported this subflow as broken. Thus, I would put pre_established in the definite set as well.

>   • TCPF_CLOSE_WAIT && tcp_sk(sk)->mptcp->pre_established or TCPF_CLOSE_WAIT ?? (Doubt! according to mptcp_sk_can_send function it is possible to send data in this state if no in pre_established MPTCP state. We remove it? or maintain it to avoid data sending in this state? def or temp unavailability?)

A subflow cannot be in pre-established state and close-wait at the same time. Because, this would mean that the host has received a subflow-FIN which moved it out of pre-established state (see mptcp_data_ack()).

To summarize, we can send data only if tcp_passive_open is true, or the subflow is not pre_established _and_ either in established or close-wait state.
All other states should be definitively non-eligible.

Cheers,
Christoph

**Comment by ghost** [comment]
> To summarize, we can send data only if tcp_passive_open is true, or the subflow is not pre_established _and_ either in established or close-wait state. All other states should be definitively non-eligible.

So, taking into consideration your comments it is quite easy to split the current mptcp_is_available function in two functions. The first lines of mptcp_is_available would be the "definitive_unavailable" since in this state we check the proper TCP state and MPTCP states (pre_establish, ...) of each subflow:

``` C
/* Set of states for which we are allowed to send data */
        if (!mptcp_sk_can_send(sk))
                return false;

        /* We do not send data on this subflow unless it is
         * fully established, i.e. the 4th ack has been received.
         */
        if (tp->mptcp->pre_established)
                return false;

        if (tp->pf)
                return false;
```

Whereas the rest of the code of mptcp_is_available, more related to window congestion and segments' order, would be the temporally unavailable state.

Quite straightforward.

**Comment by ghost** [comment]
> Then we would be continuously reinjecting segments on used sk's when we the pipes (congestion windows) of the other sk's are all filled.

Thanks for your explanation about reinjected packets over used sockets. It was really illuminating.

However, one point that we should consider is what we do when the skb has been sent by all the available main and backup sks. What should be do? We have two options: to resend the packet again through one already sent sk or discard quietly the skb from the reinjection queue and pass to the process the following skb. In the current implementation, the skb is sent again a second time through an already sk and the path_mask is cleared. So, to get this we need that get_subflow_from_selectors function provides us an used sk instead of NULL sk pointer and to flag it as used skb. 

So, I will modify the function as follows in order to get:
- always return the calling function one unused or at least one used sk.
- flag the calling function 
  - if returning sk is valid -> is it used or unused?
  - if returning sk is NULL -> are there temporally unavailable sks?
  
  ``` C
  static struct sock *get_subflow_from_selectors(struct mptcp_cb *mpcb, 
                         struct sk_buff *skb, 
                         bool (*main_selector)(struct tcp_sock *, struct sk_buff *),
                         bool zero_wnd_test, 
                         bool *force)
  { 
  cnt_temp_una = 0
  bool found_unused = false;
  bestsk = NULL;
  
  mptcp_for_each_sk (mpcb, sk)
  { 
    bool unused = false; 
    tp = tcp_sk(sk); 
    /* First, we choose among the main-set - whether this is a low-prio subflow or not */ 
    if (!(*main_selector)(tp, skb)) 
      continue;
  
    if (!mptcp_dont_reinject_skb())
      unused = true;
    else if (found_unused)
      /* If this is a reinjection, and we already had an unused subflow that 
      * was eligible, we continue - no need to check reinjection-subflows anymore. */ 
      continue;
  
    if (mptcp_is_def_unavailable(sk))
      continue;
  
    if (mptcp_is_temp_unavailable(sk, skb, zero_wnd_test)) {
      cnt_temp_una++;
      continue;
    }
  
    if (unused) { 
      /* It's the first time we encounter an unused 
      * subflow - thus we reset the bestsk (which might 
      * have been set to a reinjection-subflow. */ 
    if (!found_unused) {
          min_srtt = 0xffffffff;
      bestsk = NULL;
        }
    found_unused = true; 
    }
  
    if (tp->srtt_us < main_min_srtt) { 
      min_srtt = tp->srtt_us; 
      bestsk = sk;
    } 
  } 
  
  if (bestsk) {
    /* bestsk is valid not NULL
     * The flag is used to differenciate between previously used or not-used sks */
    if (found_unused)
      *force = true;
    else
      *force = false;
  } else {
    /* bestsk is NULL
     * The flas is used to indicate if there are temporally unavailable sks */
    if (cnt_temp_una)
      *force = true;
    else
      *force = false;
  }
  
  return bestsk;
  }
  ```

The call to the function would be:

``` C
sk = get_subflow_from_selectors(..., &main_selector,...)
if (force)
  /* we return a valid sk if it is unused or a null sk if there are temp unavailable sks */
  return sk;

sk = get_subflow_from_selectors(..., &backup_selector,...)
if (sk && !force)
  /* we return used backup sk, so reset path_mask */
  TCP_SKB_CB(skb)->path_mask = 0;
/* we always return the sk from backup_selector: it could be an unused, used or NULL sk */
return sk;
```
